KIPs/kip-114.md

```diff
@@ -0,0 +1,379 @@
+- (Dependability) `mixHash` only relies on the validators.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+The terms hash_to_point (hereinafter referred to as hash), public key, proof-of-possession (hereinafter referred to as pop) are from the ciphersuite [BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_POP\_](https://www.ietf.org/archive/id/draft-irtf-cfrg-bls-signature-05.html#section-4.2.3).
+
+### Parameters
+
+| Constant     | Value |
+| ------------ | ----- |
+| `FORK_BLOCK` | TBD   |
+
+### Block Structure
+
+Beginning with `FORK_BLOCK`, client software must set the fields in a block header:
+
+- `randomReveal`: 96 bytes. It must be specified in the 15th field (0-indexed) of a block header.
+
+- `mixHash`: 32 bytes. It must be specified in the 16th field (0-indexed) of a block header.
+
+### Block Processing
+
+#### Block Proposal
+
+Beginning with `FORK_BLOCK`, the proposer must fill the new fields in the block proposal:
+
+- `randomReveal`: The proposer must sign the hash of the proposing block number.
+
+- `mixHash`: The proposer must XOR the keccak256 hash of `randomReveal` and the previous block's `mixHash`. If the previous block's `mixHash` is null (e.g., previous block is before `FORK_BLOCK`), then use 32 bytes of zero instead.
+
+#### Validation
+
+Beginning with `FORK_BLOCK`, validators must validate the new fields:
+
+- `randomReveal`
+
+  - The validators must fetch the public key and the pop of the proposer by `getInfo(cnNodeId)` of the BLS registry standardized by [KIP-113](https://github.com/klaytn/kips/blob/main/KIPs/kip-113.md#validation).
+  - The validators must verify the validity of the proposer public key using pop as mentioned in [here](https://github.com/klaytn/kips/blob/main/KIPs/kip-113.md#validation).
+  - The validators must verify if `randomReveal` is signed by the proposer public key.
+
+- `mixHash`: The validators must verify the validity of the proposed `mixHash` by calculating `mixHash` using the aforementioned `mixHash` algorithm.
+
+### Pseudocode
+
+See the following pseudocode that describes the block processing in python:
+
+```python
+DEFAULT_MIXHASH = b"\x00" * 32
+
+def fill_kip114_header(newHeader, prevHeader, privateKey):
+    newHeader.randomReveal = calc_random_reveal(privateKey, newHeader.number)
+
+    if newHeader.number == FORK_BLOCK:
+        prevMixHash =  DEFAULT_MIXHASH
+    else:
+        prevMixHash = prevHeader.mixHash
+    newHeader.mixHash = calc_mix_hash(prevMixHash, newHeader.randomReveal)
+
+def calc_random_reveal(privateKey, headerNumber):
+    return sign(privateKey, headerNumber)
+
+def calc_mix_hash(prevMixHash, randomReveal):
+    return xor(prevMixHash, keccak256(randomReveal))
+
+def verify_kip114_header(header, prevHeader):
+    [proposerPubkey, proposerPop] = getProposerPubkeyPop()
```

Comment:
```suggestion
    [proposerPubkey, proposerPop] = get_proposer_pubkey_pop()
```

I think it would be good to keep using the same format for function names.

Reply:
also where the value of pop comes from? in the block header or smart contract registry?

Reply:
Oh, I found the answer here.
https://github.com/klaytn/kips/pull/144/files#diff-9febbc100dfaf183d5452fc7b52cc56be7d202366db2d11e35d3081bb7f88048R333